### DISCUSSION TITLE
Add Goodix fingerprint device firmware update

### DISCUSSION THREAD

**Discussion by mpi3d** [fwupd/fwupd#3637] [Q&A]
Hello, I'm currently developing drivers for Goodix fingerprint chips with a community.

The chips use the TLS protocol to communicate. So we need to write a PSK key.
But to do that we need to erase the chip firmware, then it reboots in a special mode.
In this mode, we need to write the PSK and the firmware.
After that, we are able to communicate with the device.

We can't read the PSK from the device, so we need to write it each time that the computer booted another OS (Windows). This is very problematic for dual boot.

We have already successfully flashed the chip firmware. And we have already retrieved an image from the sensor.

So we began writing the driver.
But libfprint developers don't want to include a firmware flashing in the driver.

So we separated both:
 + Driver: [rootd/libfprint](https://github.com/rootd/libfprint)
 + Firmware flashing: [mpi3d/goodix-fp-dump](https://github.com/mpi3d/goodix-fp-dump)

But many people in the community asked to add the firmware flashing in fwupd. So that why I'm opening this issue.

Our last problem is that firmware binaries are the Goodix property, so it's not legal to distribute them.
Since I'm not a lawyer, I don't know how to distribute them in a legal way.
But I hope you have some ideas of how to distribute that legally.

The goodix TLS community

**Comment by superm1** [comment]
>We can't read the PSK from the device, so we need to write it each time that the computer booted another OS (Windows). This is very problematic for dual boot.

This makes sense.  I think I would treat the dual boot problem and the firmware flashing problem as two tangential but related issues.  So really this breaks down into 3 questions:
1) The dual boot problem is really an availability of the PSK issue.  
2) Firmware flashing (fwupd support)
3) Firmware distribution

I suggest get the firmware flashing sorted first, the distribution next (or in parallel) and then we can figure out some ways that the PSK can be shared between multiple OSes.

>But many people in the community asked to add the firmware flashing in fwupd. So that why I'm opening this issue.

Makes sense.  So we have a plugin writing tutorial available here:
https://fwupd.github.io/libfwupdplugin/tutorial.html

We have a contributing guide here:
https://github.com/fwupd/fwupd/blob/master/CONTRIBUTING.md

You can use those to port your standalone flashing support into fwupd.  
You can give it a try and open a draft pull request and if you have questions along the way you can leave some comments in the PR and we can try to help you with them.

>We can't read the PSK from the device, so we need to write it each time that the computer booted another OS (Windows). This is very problematic for dual boot.

How is this PSK used and created?  When initially shipped it's empty, and then the OS driver creates it the first time that it communicates with the device?  If so, would you consider modifying the driver to write it somewhere easily accessible to multiple operating systems but that needs permissions to access like the EFI NVRAM or the TPM?

>Our last problem is that firmware binaries are the Goodix property, so it's not legal to distribute them.
Since I'm not a lawyer, I don't know how to distribute them in a legal way.
But I hope you have some ideas of how to distribute that legally.

Firmware binaries used by fwupd is most commonly distributed by the [Linux Vendor Firmware Service](https://fwupd.org/).  You don't need to distribute the source for the binaries, just the binaries and metadata associated with them to describe the firmware such as device it's for, version, any changelog you want to share.  When you distribute them via LVFS all end users running Linux distributions with fwupd and your device will be able to download the updates.

You can see here a list of all the vendors that use this service: https://fwupd.org/lvfs/vendors/

FYI, it may take a few weeks to get an account made as the person who normally does this is on PTO.

**Reply by mpi3d** [reply]
Ok, thank you for the answer.

I will work on the libfprint driver before (Because it's not released yet)
They should also have a possibility of reusing some libfprint driver code.
For now, the only working driver is written in Python.

The PSK is 32 bytes. But all the PSK stuff is encrypted.
This mean that to write the PSK we send an encrypted payload.
We already worked hard to get the corresponding encryption for an all 0 PSK. But that the only encrypted PSK that we know.
The PSK is stored on the OS file system, not in the EFI or anything that is shared between the OSes.
And more, we don't know where to find this PSK on Windows. And if we find it, we still need to decrypt it.
So there is no way to get the PSK and to avoid reflashing the firmware.
For dual boot between Linux and Linux (or even FreeBSD or anything that use the libfprint driver) there is no need to reflash the firmware. Since we will always write an all 0 PSK. So there will be no need to alter the PSK.
The dual boot problem is only concerning Windows.

If it is legal to distribute the firmware binaries, then I'm asking for an account at LVFS.

**Reply by superm1** [reply]
>I will work on the libfprint driver before (Because it's not released yet)
They should also have a possibility of reusing some libfprint driver code.
For now, the only working driver is written in Python.

OK, happy to review when you're ready with the split of the driver and firmware flashing for fwupd.

>The PSK is 32 bytes. But all the PSK stuff is encrypted.
This mean that to write the PSK we send an encrypted payload.
We already worked hard to get the corresponding encryption for an all 0 PSK. But that the only encrypted PSK that we know.
The PSK is stored on the OS file system, not in the EFI or anything that is shared between the OSes.
And more, we don't know where to find this PSK on Windows. And if we find it, we still need to decrypt it.

Oh I didn't realize you don't work for Goodix...
It would be good to bring them into this as well.

>If it is legal to distribute the firmware binaries, then I'm asking for an account at LVFS.

Well it's legal for Goodix to distribute them.  If you don't (legally) represent them, it will be up to them to do this task.
This shouldn't block the fwupd work, you can still get the updater working and integrated.  End users can install it using
`fwupdtool install-blob foo.bin` until Goodix publishes their own CAB to LVFS.

**Reply by mpi3d** [reply]
Ok no problem.

**Reply by mpi3d** [reply]
@superm1 Just: Do you have any idea of the best way to distribute the firmwares?

**Reply by superm1** [reply]
I guess that opens the question - is this custom open source firmware or proprietary firmware?

The best way to distribute proprietary firmware is for Goodix to upload them to LVFS, and the whole ecosystem will benefit.
Without their permission the best you can do is offer a script somewhere (such as in `./contrib/`) that will extract the firmware from an executable that user downloads from a support site and converts it into the appropriate format.

**Reply by mpi3d** [reply]
I got the firmware by sniffing the USB traffic between the Goodix driver and the device. So I don't know what are my right. I'm not a lawyer.

**Reply by superm1** [reply]
Okay so it's proprietary firmware, you will not be able to redistribute it to LVFS.

**Reply by mpi3d** [reply]
Ok

**Reply by mpi3d** [reply]
@superm1 Btw, it seems that fwupd already support [goodix-moc](https://github.com/fwupd/fwupd/tree/master/plugins/goodix-moc) (it would be goodix-tls for our firmwares). But where are the goodix-moc firmwares?

**Reply by hughsie** [reply]
> But where are the goodix-moc firmwares?

It's currently uploaded but embargoed. I think the Goodix team might be open to the idea of uploading more.

**Reply by mpi3d** [reply]
Ok. So I think the best idea is to write the driver for fwupd then ask Goodix to publish the firmware binaries.
